src/stash/src/upgrade/json_to_proto.rs

```diff
@@ -0,0 +1,1381 @@
+            }
+            COLLECTION_STORAGE_SHARD_FINALIZATION => {
+                migrate_collection::<ShardId, (), objects_v15::ShardId, ()>(client, id)
+                    .await
+                    .context("migrating storage-shards-to-finalize")?;
+            }
+            collection_name => {
+                // Complain loudly, but don't error out the entire migration. All of the
+                // collections listed above are the ones we know about, and are currently using.
+                // If there exists more it doesn't really matter since nothing in code references
+                // them today, but we want to know if there are.
+                tracing::error!("Unrecognized collection name {collection_name}");
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Migrates a single collection, based on an `id` from the JSON types `(KJ, VJ)` to the protobuf
+/// types `(KP, VP)`.
+async fn migrate_collection<KJ, VJ, KP, VP>(
+    client: &'_ Transaction<'_>,
+    collection_id: i64,
+) -> Result<(), anyhow::Error>
+where
+    KJ: for<'de> Deserialize<'de>,
+    VJ: for<'de> Deserialize<'de>,
+    KP: ::prost::Message + From<KJ>,
+    VP: ::prost::Message + From<VJ>,
+{
+    // Get all of the rows for a single collection and convert them to proto.
+    let mut proto_rows = client
+        .query(
+            "SELECT key, value, time, diff FROM data WHERE collection_id = $1",
+            &[&collection_id],
+        )
+        .await
+        .context("getting rows")?
+        .into_iter()
+        .map(|row| {
+            let key: serde_json::Value = row.get("key");
+            let val: serde_json::Value = row.get("value");
+            let time: i64 = row.get("time");
+            let diff: i64 = row.get("diff");
+
+            let key_json: KJ = serde_json::from_value(key).expect("must deserialize");
+            let val_json: VJ = serde_json::from_value(val).expect("must deserialize");
+
+            let key_proto = KP::from(key_json);
+            let val_proto = VP::from(val_json);
+
+            ((key_proto, val_proto), time, diff)
+        });
+
+    // Insert all of the converted protos into the new table.
+    loop {
+        let mut batch = Vec::new();
+        let mut rows = 0;
+        let mut serialized_size = 0;
+
+        // Keep popping rows into the batch.
+        while let Some(((key, val), time, diff)) = proto_rows.next() {
+            serialized_size += key.encoded_len();
+            serialized_size += val.encoded_len();
+
+            batch.push(((key.encode_to_vec(), val.encode_to_vec()), time, diff));
+            rows += 1;
+
+            // Once our serialized size is greater than our max, break to run an insert.
+            if serialized_size > INSERT_BATCH_MAX_SIZE {
+                break;
+            }
+        }
+
+        // We must have removed all of the rows, exit!
+        if rows == 0 {
+            return Ok(());
+        }
```

Comment:
@mjibson batching inserts based on a max size is happening here, which should prevent the issue fixed in #18916

Reply:
Ah didn't see this! Great.